service/history/visibilityQueueTaskExecutor.go

```diff
@@ -477,6 +481,31 @@ func (t *visibilityQueueTaskExecutor) isCloseExecutionVisibilityTaskPending(task
 	return !queues.IsTaskAcked(queryTask, visibilityQueueState)
 }
 
+// cleanupExecutionInfo cleans up workflow execution info after visibility close
+// task has been processed and acked by visibility store.
+func (t *visibilityQueueTaskExecutor) cleanupExecutionInfo(
+	ctx context.Context,
+	task *tasks.CloseExecutionVisibilityTask,
+) (retError error) {
+	weContext, release, err := getWorkflowExecutionContextForTask(ctx, t.cache, task)
+	if err != nil {
+		return err
+	}
+	defer func() { release(retError) }()
+
+	mutableState, err := weContext.LoadMutableState(ctx)
```

Comment:
Since the lock is released before locked again here I think we need to redo the checks on workflow state and workflow version as conflict resolution etc can happen once the lock is released and rebuild mutable state. cc @yux0

Reply:
Updated to do the checks.